File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Database/DatabaseRecord.cs
// <copyright file="RemoteDatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Collections.Generic;
    using Meta;
    using System.Linq;
    using Allors.Protocol.Json.Api.Sync;
    using Collections;

    internal class DatabaseRecord : IRecord
    {
        private Permission[] deniedPermissions;
        private AccessControl[] accessControls;

        private Dictionary<IRelationType, object> roleByRelationType;
        private SyncResponseRole[] syncResponseRoles;

        private readonly Database database;

        internal DatabaseRecord(Database database, long id, IClass @class)
        {
            this.database = database;
            this.Id = id;
            this.Class = @class;
            this.Version = 0;
        }

        internal DatabaseRecord(Database database, ResponseContext ctx, SyncResponseObject syncResponseObject)
        {
            this.database = database;
            this.Id = syncResponseObject.Id;
            this.Class = (IClass)this.database.MetaPopulation.FindByTag(syncResponseObject.ObjectType);
            this.Version = syncResponseObject.Version;
            this.syncResponseRoles = syncResponseObject.Roles;
            this.AccessControlIds = syncResponseObject.AccessControls != null ? (ISet<long>)new HashSet<long>(ctx.CheckForMissingAccessControls(syncResponseObject.AccessControls)) : EmptySet<long>.Instance;
            this.DeniedPermissionIds = syncResponseObject.DeniedPermissions != null ? (ISet<long>)new HashSet<long>(ctx.CheckForMissingPermissions(syncResponseObject.DeniedPermissions)): EmptySet<long>.Instance;
        }

        internal IClass Class { get; }

        internal long Id { get; }

        public long Version { get; private set; }

        internal ISet<long> AccessControlIds { get; }

        internal ISet<long> DeniedPermissionIds { get; private set; }

        private Dictionary<IRelationType, object> RoleByRelationType
        {
            get
            {
                if (this.syncResponseRoles != null)
                {
                    var meta = this.database.MetaPopulation;

                    var metaPopulation = this.database.MetaPopulation;
                    this.roleByRelationType = this.syncResponseRoles.ToDictionary(
                        v => (IRelationType)meta.FindByTag(v.RoleType),
                        v =>
                        {
                            var roleType = ((IRelationType)metaPopulation.FindByTag(v.RoleType)).RoleType;

                            var objectType = roleType.ObjectType;
                            if (objectType.IsUnit)
                            {
                                return UnitConvert.FromJson(roleType.ObjectType.Tag, v.Value);
                            }

                            if (roleType.IsOne)
                            {
                                return v.Object;
                            }

                            return v.Collection;
                        });

                    this.syncResponseRoles = null;
                }

                return this.roleByRelationType;
            }
        }

        private AccessControl[] AccessControls =>
            this.accessControls = this.accessControls switch
            {
                null when this.AccessControlIds == null => Array.Empty<AccessControl>(),
                null => this.AccessControlIds
                    .Select(v => this.database.AccessControlById[v])
                    .ToArray(),
                _ => this.accessControls
            };

        private Permission[] DeniedPermissions =>
            this.deniedPermissions = this.deniedPermissions switch
            {
                null when this.DeniedPermissionIds == null => Array.Empty<Permission>(),
                null => this.DeniedPermissionIds
                    .Select(v => this.database.PermissionById[v])
                    .ToArray(),
                _ => this.deniedPermissions
            };

        public object GetRole(IRoleType roleType)
        {
            object @object = null;
            _ = this.RoleByRelationType?.TryGetValue(roleType.RelationType, out @object);
            return @object;
        }

        public bool IsPermitted(Permission permission) =>
            permission != null &&
            !this.DeniedPermissions.Contains(permission) &&
            this.AccessControls.Any(v => v.PermissionIds.Any(w => w == permission.Id));

        internal void UpdateDeniedPermissions(long[] updatedDeniedPermissions)
        {
            if (this.deniedPermissions == null)
            {
                this.DeniedPermissionIds = EmptySet<long>.Instance;
            }
            else
            {
                if (!this.DeniedPermissionIds.SetEquals(updatedDeniedPermissions))
                {
                    this.DeniedPermissionIds = new HashSet<long>(updatedDeniedPermissions);
                }
            }
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/database/DatabaseRecord.ts
import { Class, RelationType, RoleType } from '@allors/workspace/meta/system';
import { DatabaseRecord as SystemDatabaseRecord, IRange } from '@allors/workspace/adapters/system';
import { SyncResponseObject, SyncResponseRole } from '@allors/protocol/json/system';
import { DatabaseConnection } from './DatabaseConnection';
import { ResponseContext } from './Security/ResponseContext';
import { unitFromJson } from '../json/fromJson';

export class DatabaseRecord extends SystemDatabaseRecord {
  grants: IRange<number>;
  revocations: IRange<number>;

  private _roleByRelationType?: Map<RelationType, unknown>;
  private syncResponseRoles?: SyncResponseRole[];

  constructor(public readonly database: DatabaseConnection, cls: Class, id: number, public version: number) {
    super(cls, id, version);
  }

  static fromResponse(database: DatabaseConnection, ctx: ResponseContext, syncResponseObject: SyncResponseObject): DatabaseRecord {
    const obj = new DatabaseRecord(database, database.configuration.metaPopulation.metaObjectByTag.get(syncResponseObject.c) as Class, syncResponseObject.i, syncResponseObject.v);
    obj.syncResponseRoles = syncResponseObject.ro;
    obj.grants = ctx.checkForMissingGrants(syncResponseObject.g);
    obj.revocations = ctx.checkForMissingRevocations(syncResponseObject.r);
    return obj;
  }

  get roleByRelationType(): Map<RelationType, unknown> {
    if (this.syncResponseRoles != null) {
      const metaPopulation = this.database.configuration.metaPopulation;
      this._roleByRelationType = new Map(
        this.syncResponseRoles.map((v) => {
          const relationType = metaPopulation.metaObjectByTag.get(v.t) as RelationType;
          const roleType = relationType.roleType;
          const objectType = roleType.objectType;

          let role: unknown;

          if (objectType.isUnit) {
            role = unitFromJson(objectType.tag, v.v);
          } else {
            if (roleType.isOne) {
              role = v.o;
            } else {
              role = v.c;
            }
          }

          return [relationType, role];
        })
      );

      delete this.syncResponseRoles;
    }

    return this._roleByRelationType;
  }

  getRole(roleType: RoleType): unknown {
    return this.roleByRelationType?.get(roleType.relationType);
  }

  isPermitted(permission: number): boolean {
    if (this.grants == null) {
      return false;
    }

    if (this.database.ranges.has(this.revocations, permission)) {
      return false;
    }

    if (this.grants == null) {
      return false;
    }

    return this.grants.some((v) => this.database.ranges.has(this.database.grantById.get(v).permissionIds, permission));
  }
}
